File: lab3/gnu/galois_field.cpp
#include "galois_field.hpp"

#include <tuple>



namespace gf {

value_type value(value_type a, order_type order) {
    return a % order;
}

value_type add(value_type a, value_type b, order_type order) {
    return (a + b) % order;
}

value_type subtract(value_type a, value_type b, order_type order) {
    return (order + (a - b) % order) % order;
}

value_type multiply(value_type a, value_type b, order_type order) {
    return (a * b) % order;
}

value_variant_type divide(value_type a, value_type b, order_type order) {
    if (!b)
        return "Cannot invert 0 (mod " + std::to_string(order) + ")";

    const auto b_inverse = detail::inverse(b, order);
    if (!b_inverse)
        return "Cannot invert " + std::to_string(b) + " (mod " + std::to_string(order) + ")";

    return (a * b_inverse.value()) % order;
}

value_type power(value_type base, value_type exponent, order_type order) {
    value_type result = 1;

    while (exponent > 0) {
        if (exponent % 2)
            result = (result * base) % order;
        base = (base * base) % order;
        exponent /= 2;
    }

    return result;
}

namespace detail {

std::optional<value_type> inverse(value_type a, order_type order) {
    value_type t = 0, new_t = 1;
    value_type r = order, new_r = a;

    while (new_r) {
        value_type quotient = r / new_r;
        std::tie(t, new_t) = std::make_pair(new_t, t - quotient * new_t);
        std::tie(r, new_r) = std::make_pair(new_r, r - quotient * new_r);
    }

    if (r > 1)
        return std::nullopt;

    return t < 0 ? t + order : t;
}

} // namespace detail

} // namespace gf

File: lab3/python/galois_field.py
import numpy as np
from typing import Union


class GaloisField:
    class Type:
        value = np.int64
        order = np.uint32
        error = str


    @staticmethod
    def value(a: Type.value, order: Type.order) -> Type.value:
        return np.mod(a, order)


    @staticmethod
    def add(a: Type.value, b: Type.value, order: Type.order) -> Type.value:
        return np.mod(a + b, order)


    @staticmethod
    def subtract(a: Type.value, b: Type.value, order: Type.order) -> Type.value:
        return np.mod(order + np.mod(a - b, order), order)


    @staticmethod
    def multiply(a: Type.value, b: Type.value, order: Type.order) -> Type.value:
        return np.mod(a * b, order)


    @staticmethod
    def _inverse(a: Type.value, order: Type.order):
        t, new_t = GaloisField.Type.value(0), GaloisField.Type.value(1)
        r, new_r = order, a

        while new_r:
            quotient = r // new_r
            t, new_t = new_t, t - quotient * new_t
            r, new_r = new_r, r - quotient * new_r

        if r > 1:
            return None

        return t + order if t < 0 else t


    @staticmethod
    def divide(a: Type.value, b: Type.value, order: Type.order) -> Union[Type.value, Type.error]:
        if not b:
            return f"Cannot invert 0 (mod {order})"

        b_inverse = GaloisField._inverse(b, order)
        if b_inverse is None:
            return f"Cannot invert {b} (mod {order})"

        return np.mod(a * b_inverse, order)


    @staticmethod
    def power(base: Type.value, exponent: Type.value, order: Type.order) -> Type.value:
        result = GaloisField.Type.value(1)

        while exponent > 0:
            if np.mod(exponent, 2):
                result = np.mod(result * base, order)
            base = np.mod(base * base, order)
            exponent //= 2

        return result
